python/glow/gwas/log_reg.py

```diff
@@ -242,20 +282,44 @@ def _logistic_regression_inner(genotype_pdf: pd.DataFrame, log_reg_state: LogReg
     c, d: covariate
     '''
     X = np.column_stack(genotype_pdf[_VALUES_COLUMN_NAME].array)
+
+    # For approximate Firth correction, we perform a linear residualization
+    if correction == correction_approx_firth:
+        X = gwas_fx._residualize_in_place(X, Q)
+
     with oe.shared_intermediates():
         X_res = _logistic_residualize(X, C, Y_mask, log_reg_state.gamma, log_reg_state.inv_CtGammaC)
         num = gwas_fx._einsum('sgp,sp->pg', X_res, log_reg_state.Y_res)**2
         denom = gwas_fx._einsum('sgp,sgp,sp->pg', X_res, X_res, log_reg_state.gamma)
     t_values = np.ravel(num / denom)
     p_values = stats.chi2.sf(t_values, 1)
 
-    if fallback_method != fallback_none:
-        # TODO: Call approx firth here
-        ()
-
     del genotype_pdf[_VALUES_COLUMN_NAME]
     out_df = pd.concat([genotype_pdf] * log_reg_state.Y_res.shape[1])
     out_df['tvalue'] = list(np.ravel(t_values))
     out_df['pvalue'] = list(np.ravel(p_values))
     out_df['phenotype'] = phenotype_names.repeat(genotype_pdf.shape[0]).tolist()
+
+    if correction != correction_none:
+        out_df['correction_method'] = 'none'
```

Comment:
I'm still not convinced we need this field. Seems like it would be sufficient to just have the `correction_succeeded` column -- `True`, `False`, or `null` for no correction.